### PR TITLE
Feature/domain queries

### PR DESCRIPTION
**Pull request by grimadas**
## What is this pull request?
New domain specific permissions. No we can specify specific domain permissions for queries.
   
## Why do you implement it? Why do we need this pull request?
To provide more flexibility to domain permissions.
  
## How to use the features provided in the pull request?
The examples of domain permissions are demonstrated in query_execution test.

## Details/Features
List of features / major commits
- Refactor query_execution test with fixtures
- Add domain permissions
- Optimize the number of database access for getRoles, getPermissions used in queries

### CODE REVIEW

File: irohad/model/impl/query_execution.cpp [hunk 1]
```diff
@@ -283,8 +317,9 @@ iroha::model::QueryProcessingFactory::execute(
     return executeGetAccountTransactions(*qry);
   }
   if (instanceof <iroha::model::GetAccountAssetTransactions>(query.get())) {
-    auto qry = std::static_pointer_cast<
-        const iroha::model::GetAccountAssetTransactions>(query);
+    auto qry = std::
+        static_pointer_cast<const iroha::model::GetAccountAssetTransactions>(
```
Comment (l4l) [inline irohad/model/impl/query_execution.cpp]: There's:
https://github.com/hyperledger/iroha/blob/b7f988b7a21fa23e6e68be6d992eaf8a3944d957/irohad/model/impl/query_execution.cpp#L33
So no need to write full namespace

File: test/module/irohad/validation/query_execution.cpp [hunk 2]
```diff
@@ -18,322 +18,712 @@
 #include "module/irohad/ametsuchi/ametsuchi_mocks.hpp"
 
 #include <model/queries/responses/account_assets_response.hpp>
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: Fix brackets plz

File: test/module/irohad/validation/query_execution.cpp [hunk 3]
```diff
@@ -18,322 +18,712 @@
...
  */
-auto ACCOUNT_ID = "test@test";
-auto ADMIN_ID = "admin@test";
-auto DOMAIN_NAME = "test";
-auto ADVERSARY_ID = "adversary@test";
-auto ASSET_ID = "coin";
-auto ADMIN_ROLE = "admin";
+TEST_F(GetAccountTest, MyAccountValidCase) {
+  EXPECT_CALL(*wsv_query, getAccountRoles(admin_id))
+      .Times(2)
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: It's not really clear why is it being called 2 times. May you leave a comment?

File: test/module/irohad/validation/query_execution.cpp [hunk 4]
```diff
@@ -18,322 +18,712 @@
...
+      .WillRepeatedly(Return(admin_roles));
+  EXPECT_CALL(*wsv_query, getRolePermissions(admin_role))
+      .WillOnce(Return(role_permissions));
+  EXPECT_CALL(*wsv_query, getAccount(admin_id)).WillOnce(Return(creator));
+  auto response = validateAndExecute();
+  auto cast_resp =
+      std::static_pointer_cast<iroha::model::AccountResponse>(response);
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: Same with the full namespace
check plz the whole file as well

File: test/module/irohad/validation/query_execution.cpp [hunk 5]
```diff
@@ -18,322 +18,712 @@
...
-  dummy.domain_id = DOMAIN_NAME;
-  dummy.quorum = 1;
-  return dummy;
+/**
+ * @given initialized storage, granted permission
+ * @when get account information about other user
+ * @then Return error
+ */
+TEST_F(GetAccountTest, GrantAccountValidCase) {
+  get_account->account_id = account_id;
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: Maybe create an additional fixture? This tests are looking very similar

File: test/module/irohad/validation/query_execution.cpp [hunk 6]
```diff
@@ -18,322 +18,712 @@
...
+class GetAccountTransactionsTest : public QueryValidateExecuteTest {
+ public:
+  void SetUp() override {
+    QueryValidateExecuteTest::SetUp();
+    get_tx = std::make_shared<GetAccountTransactions>();
+    get_tx->account_id = admin_id;
+    get_tx->creator_account_id = admin_id;
+    query = get_tx;
+    role_permissions = {can_get_my_acc_txs};
+    N = 3;
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: I guess this declaration can be moved to the class field definition

File: test/module/irohad/validation/query_execution.cpp [hunk 7]
```diff
@@ -18,322 +18,712 @@
...
+    txs_observable = rxcpp::observable<>::iterate([this] {
+      std::vector<iroha::model::Transaction> result;
+      for (size_t i = 0; i < N; ++i) {
+        iroha::model::Transaction current;
+        current.creator_account_id = account_id;
+        current.tx_counter = i;
+        result.push_back(current);
+      }
+      return result;
+    }());
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: Define + call lambda? Maybe just remove lambda and pass the result inside?

File: test/module/irohad/validation/query_execution.cpp [hunk 8]
```diff
@@ -18,322 +18,712 @@
...
+ * @when get account assets
+ * @then Return account asset of user
+ */
+TEST_F(GetAccountTransactionsTest, MyAccountValidCase) {
+  EXPECT_CALL(*wsv_query, getAccountRoles(admin_id))
+      .WillOnce(Return(admin_roles));
+  EXPECT_CALL(*wsv_query, getRolePermissions(admin_role))
+      .WillOnce(Return(role_permissions));
+
+  txs_observable = rxcpp::observable<>::iterate([this] {
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: You may extract it as a generic fixture method

File: test/module/irohad/validation/query_execution.cpp [hunk 9]
```diff
@@ -18,322 +18,712 @@
...
 
-  auto query = std::make_shared<GetRoles>();
-  query->creator_account_id = ADMIN_ID;
-  auto response = query_proccesor.execute(query);
-  auto cast_resp = std::static_pointer_cast<RolesResponse>(response);
-  ASSERT_EQ(1, cast_resp->roles.size());
-  ASSERT_EQ(ADMIN_ROLE, cast_resp->roles.at(0));
-  // TODO: add more test cases
+  cast_resp->transactions.as_blocking().subscribe(
+      [this](auto tx) { ASSERT_EQ(admin_id, tx.creator_account_id); }, []() {});
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: [You may omit parentheses](https://ideone.com/aXe19K)

File: test/module/irohad/validation/query_execution.cpp [hunk 10]
```diff
@@ -18,322 +18,712 @@
...
+  EXPECT_CALL(*wsv_query, getRolePermissions(admin_role))
+      .WillOnce(Return(role_permissions));
+
+  EXPECT_CALL(*block_query, getAccountTransactions(get_tx->account_id))
+      .WillOnce(Return(rxcpp::observable<>::empty<Transaction>()));
+
+  auto response = validateAndExecute();
+  auto cast_resp =
+      std::static_pointer_cast<iroha::model::TransactionsResponse>(response);
+}
```
Comment (l4l) [inline test/module/irohad/validation/query_execution.cpp]: Missing newline

**Comment by l4l** [comment]
It seems that coverage significantly decreases :( Maybe add some other cases?

File: test/module/irohad/validation/query_execution.cpp [hunk 11]
```diff
@@ -17,323 +17,1099 @@
...
+  EXPECT_CALL(*wsv_query, getAccountRoles(admin_id))
+      .WillOnce(Return(admin_roles));
+  EXPECT_CALL(*wsv_query, getRolePermissions(admin_role))
+      .WillOnce(Return(role_permissions));
+  EXPECT_CALL(*wsv_query, getAccountAsset(admin_id, asset_id))
+      .WillOnce(Return(accountAsset));
+  auto response = validateAndExecute();
+  auto cast_resp = std::static_pointer_cast<AccountAssetResponse>(response);
+  ASSERT_EQ(cast_resp->acct_asset.account_id, admin_id);
+}
```
Comment (kamilsa) [inline test/module/irohad/validation/query_execution.cpp]: Here also check asset id

**Pull request merged**